Use string dependencies in the get method of the solver [TG-2607]

romainbrenguier opened the pull request:
This is based on https://github.com/diffblue/cbmc/pull/1895 which should be merged soon, so commits up to 123541f can be ignored.
The goal is that `get` uses the information of how a string is obtained with respect to other strings to compute a possible valuation instead of using the model given by the underlying solver.
This can save us several refinement step, as the valuation obtained that way has more chances of satisfying all the string constraints.

romainbrenguier committed:
Add an evaluation method to builtin functions

This will be used to concretize strings by directly computing them from
the primitives instead of using the solver, when possible.

romainbrenguier committed:
Add const node_at function in string dependencies

This is useful to find a node with the guarantee of node modifying the graph.

romainbrenguier committed:
Activate dependency computation in string solver

This is computed at the beginning of dec_solve.

romainbrenguier committed:
Linting corrections in string_refinement

romainbrenguier committed:
Use eval of builtin function in get when available

When we can find a string in the string dependencies graph and the
corresponding function can be evaluated we return that result instead of
using get_array.

romainbrenguier committed:
Check type before adding dependencies

We check that the type is that of a string before adding the dependence.

romainbrenguier committed:
Add eval function to string_dependenciest

romainbrenguier committed:
Tests for string solver get with new dependencies

romainbrenguier committed:
Fix string constraints printing

romainbrenguier committed:
Fix linting problem in string_constraint

romainbrenguier committed:
Add concat_char builtin function for strings

romainbrenguier committed:
Add a cache for eval in string dependences

This improves the performances since the model for one string can depend
on the model for two other strings, which model may have already been
computed.
The cache must be cleared every time the model changes.

romainbrenguier committed:
Declare a nodet class in string_dependencies

This make it more convenient to reuse graph function, rather than using
a common index for two tables.

romainbrenguier committed:
Make node point to builtin func they result from

This make finding the function the string results from easier.

tautschnig commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -257,7 +257,7 @@ optionalt<exprt> string_insertion_builtin_functiont::eval(
   const auto &input1_value = eval_string(input1, get_value);
   const auto &input2_value = eval_string(input2, get_value);
-  if(!input2_value.has_value() || !input1_value.has_value())
+  if(!input1_value.has_value() || !input2_value.has_value())
```
The change seems to be completely unrelated to the commit message.

tautschnig commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -757,6 +757,7 @@ decision_proceduret::resultt string_refinementt::dec_solve()
 
   // Initial try without index set
   const auto get = [this](const exprt &expr) { return this->get(expr); };
+  dependences.clean_cache();
```
I believe the plural is "dependencies". Notably, your code sometimes already uses "dependencies".

romainbrenguier replied:
my understanding is that dependences is the the plural of dependence and dependencies the plural of dependency. Anyway I'd prefer to keep only one...

tautschnig commented on regression/jbmc-strings/StringConcat/Test.java:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    public String bufferNonDetLoop2(int width, String c) {
+	if(width < 5 || c.length() > width)
```
Strange formatting.

tautschnig commented on regression/jbmc-strings/StringConcat/test_buffer_det.desc:
```diff
@@ -0,0 +1,11 @@
+check_SAT: got SAT but the model is not correct
+--
+Using the string dependences information to compute the model, refinement
+should not be needed when the execution is deterministic.
```
No end-of-line at end-of-file. (Applies elsewhere as well.)

LAJW commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -140,4 +141,203 @@ class equation_symbol_mappingt
 };
 
+/// Base class for string functions that are built in the solver
+class string_builtin_functiont
```
When creating a virtual class hierarchy, the base class destructor should be defined and marked as virtual. i.e.

```
virtual ~string_builtin_functiont() = default;
```

Otherwise, you'll end up with memory leaks.

LAJW commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -140,4 +141,203 @@ class equation_symbol_mappingt
+class string_builtin_functiont
+{
+public:
+  string_builtin_functiont() = default;
```
If this is not intended to be instantiated, you could mark this constructor as `protected`.

LAJW commented on src/solvers/refinement/string_refinement_util.h:
```diff
@@ -140,4 +141,203 @@ class equation_symbol_mappingt
+/// Keep track of dependencies between strings.
+/// Each string points to builtin_function calls on which it depends,
+/// each builtin_function points to the strings on which the result depend.
+class string_dependencest
```
Typo - `dependencies` - missing an `i`. Also, documentation could use better grammar.

tautschnig commented:
I'd appreciate if a native speaker could chime in, but my reading of https://en.wiktionary.org/wiki/dependence#English vs. https://en.wiktionary.org/wiki/dependency#English is that the latter is the appropriate technical term.

romainbrenguier committed:
Rename class to string_dependenciest

romainbrenguier committed:
Constructor and destructor of builtin functions

Mark empty constructor as protected and define destructor as virtual default.

romainbrenguier committed:
Correct insert builtin function construction

The arguments were not in the right order.
We also need to seperate construction of concatenation since it was
using insertion.

martin-cs commented:
@tautschnig I would use depend / dependency / dependencies (as is "x == f(x) depends on x and all of f's dependencies") in technical computing English.  Will review code when I've had a chance to look at the previous PR.

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -680,7 +680,7 @@ decision_proceduret::resultt string_refinementt::dec_solve()
   output_equations(debug(), equations, ns);
 
-  string_dependencest dependences;
+  string_dependenciest dependences;
```
Can you also rename the variable, please?

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -172,6 +172,39 @@ equation_symbol_mappingt::find_equations(const exprt &expr)
+  std::transform(args.begin(), args.end(), insert, [&](const exprt &e) { // NOLINT
+    if(const auto val = numeric_cast<mp_integer>(get_value(e)))
+      return *val;
+    // INVARIANT(get_value(e).id() == ID_unknown);
```
Remove commented out code

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -526,7 +521,12 @@ bool add_node(
     dependencies.add_dependency(*string_result, builtin_function_node);
+    auto &node = dependencies.get_node(*string_result);
+    node.result_from = builtin_function_node;
+
```
remove spurious blank line

peterschrammel reviewed (commented):
Looks good. There are failing tests, though.

romainbrenguier committed:
Decompose if_expr when adding dependencies

We add the dependencies on atomic string expressions, instead of the
if_expr directly.

romainbrenguier committed:
Add name method to builtin functions

This is useful to make the dot output more informative.
